src/node/structural_equal.cc

```diff
@@ -184,10 +390,18 @@ class RemapVarSEqualHandler : public SEqualReducer::Handler {
         return it->second.same_as(rhs);
       }
       if (equal_map_rhs_.count(rhs)) return false;
+
       // Run reduce check for free nodes.
-      return vtable_->SEqualReduce(lhs.get(), rhs.get(), SEqualReducer(this, map_free_vars));
+      if (first_mismatch == nullptr) {
```

Comment:
Will it be better to wrap this in a helper function with more meaningful name to tell whether it's in fast path? `first_mismatch == nullptr` itself might be interpreted as 'if there is no mismatch' if reader doesn't have a full understanding on your change.